src/gConsent/query/query.ts

```diff
@@ -0,0 +1,262 @@
+// if there are missing cases when handling the value.
+type SupportedFilterElements =
+  typeof FILTER_ELEMENTS extends Array<infer E> ? E : never;
+
+// The type assertion is okay in the context of the type guard.
+const isSupportedFilterElement = (x: unknown): x is SupportedFilterElements =>
+  FILTER_ELEMENTS.includes(x as keyof CredentialFilter);
+
+const PAGING_RELS = ["first", "last", "prev", "next"] as const;
+
+export type CredentialResult = {
+  /**
+   * Page of Access Credentials matching the query.
+   */
+  items: DatasetWithId[];
+  /**
+   * First page of query results.
+   */
+  first?: CredentialFilter;
+  /**
+   * Previous page of query results.
+   */
+  prev?: CredentialFilter;
+  /**
+   * Next page of query results.
+   */
+  next?: CredentialFilter;
+  /**
+   * Last page of query results.
+   */
+  last?: CredentialFilter;
+};
+
+function toCredentialFilter(url: URL): CredentialFilter {
+  const filter: CredentialFilter = {};
+  url.searchParams.forEach((value, key) => {
+    if (isSupportedFilterElement(key)) {
+      Object.assign(filter, { [`${key}`]: value });
+    }
+  });
+  return filter;
+}
+
+function hasItems(x: unknown): x is { items: unknown[] } {
+  const candidate = x as { items: unknown };
+  return candidate.items !== undefined && Array.isArray(candidate.items);
+}
+
+function isObjectWithId(x: unknown): x is { id: string } {
+  const candidate = x as { id: string };
+  return typeof candidate.id === "string";
+}
+
+async function toCredentialResult(
+  response: Response,
+): Promise<CredentialResult> {
+  const result: CredentialResult = { items: [] };
+  const linkHeader = response.headers.get("Link");
+  if (linkHeader !== null) {
+    const parsedLinks = LinkHeader.parse(linkHeader);
+    PAGING_RELS.forEach((rel) => {
+      const link = parsedLinks.get("rel", rel);
+      if (link.length > 1) {
+        throw Error(
+          `Unexpected response, found more than one ${rel} Link headers.`,
+        );
+      }
+      if (link.length === 0) {
+        return;
+      }
+      result[rel] = toCredentialFilter(new URL(link[0].uri));
+    });
+  }
+  const body = await response.json();
+  if (!hasItems(body) || !body.items.every(isObjectWithId)) {
+    throw Error(
+      `Unexpected response, no items found in ${JSON.stringify(body)}`,
+    );
+  }
+  result.items = await Promise.all(
+    body.items.map((vc) => {
+      return verifiableCredentialToDataset(vc);
+    }),
+  );
+  return result;
+}
+
+function toQueryUrl(endpoint: URL, filter: CredentialFilter): URL {
+  const result = new URL(endpoint);
+  Object.entries(filter).forEach(([key, value]) => {
+    if (isSupportedFilterElement(key)) {
+      result.searchParams.append(key, value.toString());
+    }
+  });
+  return result;
+}
+
+/**
+ * Query for Access Requests or Access Grants based on a given filter.
+ *
+ * @param filter the query filter
+ * @param options query options
```

Comment:
```suggestion
 * @param filter The query filter
 * @param options Query options
```
We mostly seem to capitalize this but are not consistent about punctuation but I on't know if we actually have a standard.